### PR TITLE
unix,tty: reopen /dev/tty for stdio fds

### PR DESCRIPTION
**Pull request by saghul**
Refs: https://github.com/nodejs/node/issues/6456#issuecomment-219974514

R= @bnoordhuis, @Gottox

### CODE REVIEW

File: src/unix/tty.c [hunk 1]
```diff
@@ -81,15 +81,18 @@ int uv_tty_init(uv_loop_t* loop, uv_tty_t* tty, int fd, int readable) {
...
   if (type == UV_TTY) {
-    /* Reopening a pty in master mode won't work either because the reopened
-     * pty will be in slave mode (*BSD) or reopening will allocate a new
-     * master/slave pair (Linux). Therefore check if the fd points to a
-     * slave device.
-     */
-    if (uv__tty_is_slave(fd) && ttyname_r(fd, path, sizeof(path)) == 0)
+    if (fd <= STDERR_FILENO) {
+      r = uv__open_cloexec("/dev/tty", O_RDWR);
+    } else if (uv__tty_is_slave(fd) && ttyname_r(fd, path, sizeof(path)) == 0) {
```
Comment (txdv) [inline src/unix/tty.c]: Can we rename slave to follower so we won't get another shitstorm?

Comment (saghul) [inline src/unix/tty.c]: Andrius, I'm not really in the mood for this crap. We use the terminology in `man 4 pty`. End of the discussion.

Comment (txdv) [inline src/unix/tty.c]: It was merely a suggestion.

**Comment by Gottox** [comment]
This will break writing to other ttys again:
```
node test.js > /dev/tty0
```
I'm okay with that. Just be aware, that this adds another layer of unexpected behavior.

**Comment by saghul** [comment]
Damn, we can't win here, can we? 😭

**Comment by Gottox** [comment]
I think this fix is a good tradeoff. At least my usecases are fixed with libuv for now and you cannot break software in production. (Sorry, that I did)

**Comment by saghul** [comment]
@Gottox It's not your fault in any way! OSX is really weird, /dev/tty cannot be used with kqueue whereas /dev/ttys* can, which creates some interesting side-effects.

I'm not sure we can get away without any tradeoff, but this one seems reasonable to me. Maybe I should mention it in the docs too. @libuv/collaborators WDYT?

**Comment by saghul** [comment]
This is an ugly hack. I think I caught a bug which was hiding in plain sight: https://github.com/libuv/libuv/pull/879 so basically any tty which was opened using the select() trick resulted in blocking writes, even if we could do them async anyway.

With #879 the beahvior is consistent and we retain tty redirection working. I'd still like some more feedback on this whole thing, I might not be seeing the forest because of the threes.

**Comment by kzc** [comment]
@saghul Do you intend to close this PR due to the findings in this discussion?
https://github.com/nodejs/node/issues/6456#issuecomment-220018822

**Comment by saghul** [comment]
Yep, let's close it now.

**Pull request closed without merging**